include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_timer_tree_h
+#define dealii_timer_tree_h
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/timer.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A class for wall-time measurements enabling a modular coupling of modules in
+ * a minimally invasive way.
+ */
+class TimerTree
+{
+public:
+  /**
+   * Constructor.
+   */
+  TimerTree()
+    : id("")
+  {}
+
+  /**
+   * Clears the content of this tree. Sub trees inserted
+   * into this tree via pointers to external trees are not touched.
+   */
+  void
+  clear()
+  {
+    this->id = "";
+    data     = nullptr;
+    sub_trees.clear();
+  }
+
+  /**
+   * Inserts a measured @p wall_time into the tree, by
+   * either creating a new entry in the tree if this function is called
+   * the first time with the ID @p ids, or by adding the @p wall_time to an
+   * entry already existing in the tree.
+   */
+  void
+  insert(std::vector<std::string> const ids, double const wall_time)
+  {
+    AssertThrow(ids.size() > 0, ExcMessage("empty name."));
+
+    if (this->id == "") // the tree is currently empty
+      {
+        AssertThrow(sub_trees.empty(),
+                    ExcMessage("invalid state found. aborting."));
+
+        this->id = ids[0];
+
+        if (ids.size() == 1) // leaves of tree reached, insert the data
+          {
+            data.reset(new Data());
+            data->wall_time += wall_time;
+
+            return;
+          }
+        else // go deeper
+          {
+            std::vector<std::string> remaining_id = erase_first(ids);
+
+            std::shared_ptr<TimerTree> new_tree;
+            new_tree.reset(new TimerTree());
```

Comment:
same here and maybe also in other places; we try to avoid bare 'operator new'  calls.